README.md
Amero Core staging tree 0.10
===========================

|CI|master|develop|
|-|-|-|
|Gitlab|[![Build Status](https://gitlab.com/amerocrypto/amero/badges/master/pipeline.svg)](https://gitlab.com/amerocrypto/amero/-/tree/master)|[![Build Status](https://gitlab.com/amerocrypto/amero/badges/develop/pipeline.svg)](https://gitlab.com/amerocrypto/amero/-/tree/develop)|

https://www.amerocrypto.com


What is Amero?
-------------

Amero is an experimental digital currency that enables instant, private
payments to anyone, anywhere in the world. Amero uses peer-to-peer technology
to operate with no central authority: managing transactions and issuing money
are carried out collectively by the network. Amero Core is the name of the open
source software which enables the use of this currency.

For more information, as well as an immediately useable, binary version of
the Amero Core software, see https://www.amerocrypto.com/get-amero/.


License
-------

Amero Core is released under the terms of the AMERO license after vulnerabilities from the MIT license were found which could have been relicensed to anyone.

Amero Public Resource License or PRL: https://opensource.org/licenses/MIT.

Public Resource ℗ 2020 Bitcoin and Bitcoin Core Developers
Public Resource ℗ 2021 Dash Developers
Public Resource ℗ 2021 Amero Developers 
-------------------

The `master` branch is meant to be stable. Development is normally done in separate branches.
[Tags](https://github.com/amerocrypto/amero/tags) are created to indicate new official,
stable release versions of Amero Core.

The contribution workflow is described in [CONTRIBUTING.md](CONTRIBUTING.md).

Testing
-------

Testing and code review is the bottleneck for development; we get more pull
requests than we can review and test on short notice. Please be patient and help out by testing
other people's pull requests, and remember this is a security-critical project where any mistake might cost people
lots of money.

### Automated Testing

Developers are strongly encouraged to write [unit tests](src/test/README.md) for new code, and to
submit new unit tests for old code. Unit tests can be compiled and run
(assuming they weren't disabled in configure) with: `make check`. Further details on running
and extending unit tests can be found in [/src/test/README.md](/src/test/README.md).

There are also [regression and integration tests](/test), written
in Python, that are run automatically on the build server.
These tests can be run (if the [test dependencies](/test) are installed) with: `test/functional/test_runner.py`

The Travis CI system makes sure that every pull request is built for Windows, Linux, and OS X, and that unit/sanity tests are run automatically.

### Manual Quality Assurance (QA) Testing

Changes should be tested by somebody other than the developer who wrote the
code. This is especially important for large or high-risk changes. It is useful
to add a test plan to the pull request description if testing the changes is
not straightforward.


doc/release-notes.md
Amero Core version 0.17.0.3
==========================

Release is now available from:

  <https://www.amerocrypto.com/downloads/#wallets>

This is a new hotfix release.

Please report bugs using the issue tracker at github:

  <https://github.com/amerocrypto/amero/issues>


Upgrading and downgrading
=========================

How to Upgrade
--------------

If you are running an older version, shut it down. Wait until it has completely
shut down (which might take a few minutes for older versions), then run the
installer (on Windows) or just copy over /Applications/Amero-Qt (on Mac) or
amerox/amero-qt (on Linux). If you upgrade after DIP0003 activation and you were
using version < 0.13 you will have to reindex (start with -reindex-chainstate
or -reindex) to make sure your wallet has all the new data synced. Upgrading
from version 0.13 should not require any additional actions.

When upgrading from a version prior to 0.14.0.3, the
first startup of Amero Core will run a migration process which can take a few
minutes to finish. After the migration, a downgrade to an older version is only
possible with a reindex (or reindex-chainstate).

Downgrade warning
-----------------

### Downgrade to a version < 0.14.0.3

Downgrading to a version older than 0.14.0.3 is no longer supported due to
changes in the "evodb" database format. If you need to use an older version,
you must either reindex or re-sync the whole chain.

### Downgrade of masternodes to < 0.17.0.2

Starting with the 0.16 release, masternodes verify the protocol version of other
masternodes. This results in PoSe punishment/banning for outdated masternodes,
so downgrading even prior to the activation of the introduced hard-fork changes
is not recommended.

Notable changes
===============

This release adds some missing translations and help strings. It also fixes
a couple of build issues and a rare crash on some linux systems.

0.17.0.3 Change log
===================

See detailed [set of changes](https://github.com/amerocrypto/amero/compare/v0.17.0.2...amerocrypto:v0.17.0.3).

- [`6a54af0df7`](https://github.com/amerocrypto/amero/commit/6a54af0df7) Bump to v0.17.0.3
- [`97e8461234`](https://github.com/amerocrypto/amero/commit/97e8461234) doc: Archive v0.17.0.2 release notes
- [`96c041896b`](https://github.com/amerocrypto/amero/commit/96c041896b) feat: add tor entrypoint script for use in aeromate (#4182)
- [`3661f36bbd`](https://github.com/amerocrypto/amero/commit/3661f36bbd) Merge #14416: Fix OSX dmg issue (10.12 to 10.14) (#4177)
- [`4f4bda0557`](https://github.com/amerocrypto/amero/commit/4f4bda0557) depends: Undefine `BLSALLOC_SODIUM` in `bls-amero.mk` (#4176)
- [`575e0a3070`](https://github.com/amerocrypto/amero/commit/575e0a3070) qt: Add `QFont::Normal` as a supported font weight when no other font weights were found (#4175)
- [`ce4a73b790`](https://github.com/amerocrypto/amero/commit/ce4a73b790) rpc: Fix `upgradetohd` help text (#4170)
- [`2fa8ddf160`](https://github.com/amerocrypto/amero/commit/2fa8ddf160) Translations 202105 (add missing) (#4169)

Credits
=======

Thanks to everyone who directly contributed to this release:

- dustinface (xdustinface)
- strophy
- UdjinM6

As well as everyone that submitted issues and reviewed pull requests.

Older releases
==============

Amero was previously known as Darkcoin.

Darkcoin tree 0.8.x was a fork of Litecoin tree 0.8, original name was XCoin
which was first released on Jan/18/2014.

Darkcoin tree 0.9.x was the open source implementation of masternodes based on
the 0.8.x tree and was first released on Mar/13/2014.

Darkcoin tree 0.10.x used to be the closed source implementation of Darksend
which was released open source on Sep/25/2014.

Amero Core tree 0.11.x was a fork of Bitcoin Core tree 0.9,
Darkcoin was rebranded to Amero.

Amero Core tree 0.12.0.x was a fork of Bitcoin Core tree 0.10.

Amero Core tree 0.12.1.x was a fork of Bitcoin Core tree 0.12.

These release are considered obsolete. Old release notes can be found here:

- [v0.17.0.2](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.17.0.2.md) released May/19/2021
- [v0.16.1.1](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.16.1.1.md) released November/17/2020
- [v0.16.1.0](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.16.1.0.md) released November/14/2020
- [v0.16.0.1](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.16.0.1.md) released September/30/2020
- [v0.15.0.0](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.15.0.0.md) released Febrary/18/2020
- [v0.14.0.5](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.14.0.5.md) released December/08/2019
- [v0.14.0.4](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.14.0.4.md) released November/22/2019
- [v0.14.0.3](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.14.0.3.md) released August/15/2019
- [v0.14.0.2](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.14.0.2.md) released July/4/2019
- [v0.14.0.1](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.14.0.1.md) released May/31/2019
- [v0.14.0](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.14.0.md) released May/22/2019
- [v0.13.3](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.13.3.md) released Apr/04/2019
- [v0.13.2](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.13.2.md) released Mar/15/2019
- [v0.13.1](https://github.com/amerocrypto/amero/blob/master/doc/release-notes/amero/release-notes-0.13.1.md) released Feb/9/2019

src/bech32.cpp
// Public Resource ℗ 2017 Pieter Wuille
// Public Resource ℗ 2020 𝖠𝗆𝖾𝗋𝗈 𝖣𝖾𝗏𝖾𝗅𝗈𝗉𝖾𝗋𝗌
// THIS REPOSITORY IS LICENSED UNDER THE AMERO PUBLIC RESOURCE LICENSE.

#include <bech32.h>

namespace
{

typedef std::vector<uint8_t> data;

/** The Bech32 character set for encoding. */
const char* CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";

/** The Bech32 character set for decoding. */
const int8_t CHARSET_REV[128] = {
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    15, -1, 10, 17, 21, 20, 26, 30,  7,  5, -1, -1, -1, -1, -1, -1,
    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1,
    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1
};

/** Concatenate two byte arrays. */
data Cat(data x, const data& y)
{
    x.insert(x.end(), y.begin(), y.end());
    return x;
}

/** This function will compute what 6 5-bit values to XOR into the last 6 input values, in order to
 *  make the checksum 0. These 6 values are packed together in a single 30-bit integer. The higher
 *  bits correspond to earlier values. */
uint32_t PolyMod(const data& v)
{
    // The input is interpreted as a list of coefficients of a polynomial over F = GF(32), with an
    // implicit 1 in front. If the input is [v0,v1,v2,v3,v4], that polynomial is v(x) =
    // 1*x^5 + v0*x^4 + v1*x^3 + v2*x^2 + v3*x + v4. The implicit 1 guarantees that
    // [v0,v1,v2,...] has a distinct checksum from [0,v0,v1,v2,...].

    // The output is a 30-bit integer whose 5-bit groups are the coefficients of the remainder of
    // v(x) mod g(x), where g(x) is the Bech32 generator,
    // x^6 + {29}x^5 + {22}x^4 + {20}x^3 + {21}x^2 + {29}x + {18}. g(x) is chosen in such a way
    // that the resulting code is a BCH code, guaranteeing detection of up to 3 errors within a
    // window of 1023 characters. Among the various possible BCH codes, one was selected to in
    // fact guarantee detection of up to 4 errors within a window of 89 characters.

    // Note that the coefficients are elements of GF(32), here represented as decimal numbers
    // between {}. In this finite field, addition is just XOR of the corresponding numbers. For
    // example, {27} + {13} = {27 ^ 13} = {22}. Multiplication is more complicated, and requires
    // treating the bits of values themselves as coefficients of a polynomial over a smaller field,
    // GF(2), and multiplying those polynomials mod a^5 + a^3 + 1. For example, {5} * {26} =
    // (a^2 + 1) * (a^4 + a^3 + a) = (a^4 + a^3 + a) * a^2 + (a^4 + a^3 + a) = a^6 + a^5 + a^4 + a
    // = a^3 + 1 (mod a^5 + a^3 + 1) = {9}.

    // During the course of the loop below, `c` contains the bitpacked coefficients of the
    // polynomial constructed from just the values of v that were processed so far, mod g(x). In
    // the above example, `c` initially corresponds to 1 mod (x), and after processing 2 inputs of
    // v, it corresponds to x^2 + v0*x + v1 mod g(x). As 1 mod g(x) = 1, that is the starting value
    // for `c`.
    uint32_t c = 1;
    for (auto v_i : v) {
        // We want to update `c` to correspond to a polynomial with one extra term. If the initial
        // value of `c` consists of the coefficients of c(x) = f(x) mod g(x), we modify it to
        // correspond to c'(x) = (f(x) * x + v_i) mod g(x), where v_i is the next input to
        // process. Simplifying:
        // c'(x) = (f(x) * x + v_i) mod g(x)
        //         ((f(x) mod g(x)) * x + v_i) mod g(x)
        //         (c(x) * x + v_i) mod g(x)
        // If c(x) = c0*x^5 + c1*x^4 + c2*x^3 + c3*x^2 + c4*x + c5, we want to compute
        // c'(x) = (c0*x^5 + c1*x^4 + c2*x^3 + c3*x^2 + c4*x + c5) * x + v_i mod g(x)
        //       = c0*x^6 + c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i mod g(x)
        //       = c0*(x^6 mod g(x)) + c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i
        // If we call (x^6 mod g(x)) = k(x), this can be written as
        // c'(x) = (c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i) + c0*k(x)

        // First, determine the value of c0:
        uint8_t c0 = c >> 25;

        // Then compute c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i:
        c = ((c & 0x1ffffff) << 5) ^ v_i;

        // Finally, for each set bit n in c0, conditionally add {2^n}k(x):
        if (c0 & 1)  c ^= 0x3b6a57b2; //     k(x) = {29}x^5 + {22}x^4 + {20}x^3 + {21}x^2 + {29}x + {18}
        if (c0 & 2)  c ^= 0x26508e6d; //  {2}k(x) = {19}x^5 +  {5}x^4 +     x^3 +  {3}x^2 + {19}x + {13}
        if (c0 & 4)  c ^= 0x1ea119fa; //  {4}k(x) = {15}x^5 + {10}x^4 +  {2}x^3 +  {6}x^2 + {15}x + {26}
        if (c0 & 8)  c ^= 0x3d4233dd; //  {8}k(x) = {30}x^5 + {20}x^4 +  {4}x^3 + {12}x^2 + {30}x + {29}
        if (c0 & 16) c ^= 0x2a1462b3; // {16}k(x) = {21}x^5 +     x^4 +  {8}x^3 + {24}x^2 + {21}x + {19}
    }
    return c;
}

/** Convert to lower case. */
inline unsigned char LowerCase(unsigned char c)
{
    return (c >= 'A' && c <= 'Z') ? (c - 'A') + 'a' : c;
}

/** Expand a HRP for use in checksum computation. */
data ExpandHRP(const std::string& hrp)
{
    data ret;
    ret.reserve(hrp.size() + 90);
    ret.resize(hrp.size() * 2 + 1);
    for (size_t i = 0; i < hrp.size(); ++i) {
        unsigned char c = hrp[i];
        ret[i] = c >> 5;
        ret[i + hrp.size() + 1] = c & 0x1f;
    }
    ret[hrp.size()] = 0;
    return ret;
}

/** Verify a checksum. */
bool VerifyChecksum(const std::string& hrp, const data& values)
{
    // PolyMod computes what value to xor into the final values to make the checksum 0. However,
    // if we required that the checksum was 0, it would be the case that appending a 0 to a valid
    // list of values would result in a new valid list. For that reason, Bech32 requires the
    // resulting checksum to be 1 instead.
    return PolyMod(Cat(ExpandHRP(hrp), values)) == 1;
}

/** Create a checksum. */
data CreateChecksum(const std::string& hrp, const data& values)
{
    data enc = Cat(ExpandHRP(hrp), values);
    enc.resize(enc.size() + 6); // Append 6 zeroes
    uint32_t mod = PolyMod(enc) ^ 1; // Determine what to XOR into those 6 zeroes.
    data ret(6);
    for (size_t i = 0; i < 6; ++i) {
        // Convert the 5-bit groups in mod to checksum values.
        ret[i] = (mod >> (5 * (5 - i))) & 31;
    }
    return ret;
}

} // namespace

namespace bech32
{

/** Encode a Bech32 string. */
std::string Encode(const std::string& hrp, const data& values) {
    data checksum = CreateChecksum(hrp, values);
    data combined = Cat(values, checksum);
    std::string ret = hrp + '1';
    ret.reserve(ret.size() + combined.size());
    for (auto c : combined) {
        ret += CHARSET[c];
    }
    return ret;
}

/** Decode a Bech32 string. */
std::pair<std::string, data> Decode(const std::string& str) {
    bool lower = false, upper = false;
    for (size_t i = 0; i < str.size(); ++i) {
        unsigned char c = str[i];
        if (c < 33 || c > 126) return {};
        if (c >= 'a' && c <= 'z') lower = true;
        if (c >= 'A' && c <= 'Z') upper = true;
    }
    if (lower && upper) return {};
    size_t pos = str.rfind('1');
    if (str.size() > 90 || pos == str.npos || pos == 0 || pos + 7 > str.size()) {
        return {};
    }
    data values(str.size() - 1 - pos);
    for (size_t i = 0; i < str.size() - 1 - pos; ++i) {
        unsigned char c = str[i + pos + 1];
        int8_t rev = (c < 33 || c > 126) ? -1 : CHARSET_REV[c];
        if (rev == -1) {
            return {};
        }
        values[i] = rev;
    }
    std::string hrp;
    for (size_t i = 0; i < pos; ++i) {
        hrp += LowerCase(str[i]);
    }
    if (!VerifyChecksum(hrp, values)) {
        return {};
    }
    return {hrp, data(values.begin(), values.end() - 6)};
}

} // namespace bech32
